### PR TITLE
disallow `pointer_from_objref` on immutable values. fixes #15857

### LINKED ISSUE #15857: Dissallow pointer_from_objref on immutable values
See discussion in https://groups.google.com/forum/#!topic/julia-dev/F9PggSikPW8 that requests an issue to track this.

@vtjnash 

> this is not C – variables don't have first-class memory addresses (also, pointer_from_objref on a immutable returns a semi-random number that was correlated at some point in the past with the value).

@StefanKarpinski 

> This should really not be allowed: calling pointer_from_objref on an immutable should be an error.

### PR DESCRIPTION
**Pull request by JeffBezanson**

### CODE REVIEW

**Review by Keno — APPROVED** [review]

**Comment by vtjnash** [comment]
This seems stricter than necessary (we only might re-box isbits and unions thereof). Is that intentional for ensuring compatibility into v1.x?

File: base/refpointer.jl [hunk 1]
```diff
@@ -61,7 +61,7 @@ Ref(x::Ptr{T}, i::Integer) where {T} = x + (i - 1) * Core.sizeof(T)
 function unsafe_convert(P::Type{Ptr{T}}, b::RefValue{T}) where T
     if isbits(T) || isbitsunion(T)
         return convert(P, pointer_from_objref(b))
-    elseif _isleaftype(T)
+    elseif _isleaftype(T) && T.mutable
```
Comment (vtjnash) [inline base/refpointer.jl]: is that inferable?

File: test/core.jl [hunk 2]
```diff
@@ -5638,11 +5642,8 @@ let
     @eval @noinline constant23367(a, b) = (a ? b : $b)
     b2 = Ref(b)[] # copy b via field assignment
     b3 = B23367[b][1] # copy b via array assignment
-    @test pointer_from_objref(b) == pointer_from_objref(b)
-    @test pointer_from_objref(b) != pointer_from_objref(b2)
-    @test pointer_from_objref(b) != pointer_from_objref(b3)
-    @test pointer_from_objref(b2) != pointer_from_objref(b3)
```
Comment (vtjnash) [inline test/core.jl]: this is necessary to verify that the test is working. perhaps use the direct ccall version here?

File: base/refpointer.jl (hunk 1 above)
Comment (JeffBezanson) [reply base/refpointer.jl]: Surprisingly no, but can be fixed easily I think.

**Comment by JeffBezanson** [comment]
> This seems stricter than necessary (we only might re-box isbits and unions thereof). Is that intentional for ensuring compatibility into v1.x?

Yes, and it also seems better to make the rule as simple as possible. If this is too annoying we can relax it later.

**Pull request merged**